### PR TITLE
[ActionSheet] API Review

### PR DESCRIPTION
**Pull request by codeman7**
This is just for a review of the API. A design doc should go out ASAP to follow.

### CODE REVIEW

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: Need docs for the class as well. This should:

- [x] Describe the class' purpose.
- [x] Include a link to the Material spec.
- [x] Include a link to any similar UIKit APIs for reference.
- [x] Describe any notable differences from the UIKit APIs.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: [typo]: `or two`.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: I believe you don't have to redefine these APIs because they're part of the parent class, should be able to remove.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: typo: `will be laid out from top to bottom in the order they were added.`

Can drop the "first being top last being bottom" part.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: What happens if you change this property (and the message property) after the view controller has been presented? Document whether the client can expect the title label to be updated or not.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: [stylenit]: Indent four spaces.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: [stylenit] Here and below, there are two spaces indenting the docs when there should just be one.

Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: "An instance of MDCActionSheetAction is passed..."

The client isn't expected to pass the MDCActionSheetAction class itself.

Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: "to add an action to the action sheet".

It's not necessarily guaranteed that the actions are buttons.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: Not sure if this sentence is as-intended? Did you mean something like "Returns an action sheet action populated with the given values"?

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: [minor] `in the list` rather than `on the list`.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: Remove any reference to `cell` in the public docs - this term leaks an implementation detail that could change in the future.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: Docs for the protocol should:

- Describe that this protocol is used by the MDCActionSheetController to delegate events.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: Is this called if the user dismisses the sheet by selecting an action?

**Comment by jverkoey** [comment]
Looking good! Most feedback is just grammatical/missing docs.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (ianegordon) [reply components/ActionSheet/src/MDCActionSheetController.h]: We should match (and document) the behavior of UIAlertController in ActionSheet mode.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (ianegordon) [reply components/ActionSheet/src/MDCActionSheetController.h]: I would mark initWithCoder: and initWithNibName:bundle NS_UNAVAILABLE.  I don't expect we would support either.

**Review by ianegordon — DISMISSED** [review]

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: We will update the view to reflect the new string.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

**Comment by jverkoey** [comment]
Spoke offline with @ianegordon: am in agreement that we can mark the initWithNib/initWithCoder APIs as unavailable at this point. We can explore adding support for interface builder as a follow-up in the future.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (ianegordon) [inline components/ActionSheet/src/MDCActionSheetController.h]: Mark as designated initializer.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (ianegordon) [inline components/ActionSheet/src/MDCActionSheetController.h]: This should *not* be a designated initializer, and in the implementation should call:
[self actionSheetControllerWithTitle:title message:nil]

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (ianegordon) [inline components/ActionSheet/src/MDCActionSheetController.h]: Annotate as final to avoid subclasses.
See #2801 for an example of how to do this.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (ianegordon) [inline components/ActionSheet/src/MDCActionSheetController.h]: This is not required in the header, but in the implementation should call:
[self actionSheetControllerWithTitle:nil message:nil];

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (ajsecord) [inline components/ActionSheet/src/MDCActionSheetController.h]: "consists"

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (ajsecord) [inline components/ActionSheet/src/MDCActionSheetController.h]: "Learn more about ...", also just use "Material bottom sheet", no need for the extra "spec".

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (ajsecord) [inline components/ActionSheet/src/MDCActionSheetController.h]: Replaced "constructor" with "initializer"

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (ajsecord) [inline components/ActionSheet/src/MDCActionSheetController.h]: Replace "constructor" with "initializer"

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (ajsecord) [inline components/ActionSheet/src/MDCActionSheetController.h]: "the the"

**Review by ianegordon — APPROVED** [review]
LGTM.

**Review by ajsecord — APPROVED** [review]
LGTM with nits

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: [minorstyle] Remove the extra indentation here.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (andrewoverton) [inline components/ActionSheet/src/MDCActionSheetController.h]: The Apple version has a parameter for style, like "default", "cancel", or "destructive". Should we have our own versions of these?

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (andrewoverton) [inline components/ActionSheet/src/MDCActionSheetController.h]: Does this need to be here?

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (andrewoverton) [inline components/ActionSheet/src/MDCActionSheetController.h]: _Style nit:_ We should be consistent (at least within a single file) about whether to put newlines after and before the "/**" and "*/" doc delimiter things

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (andrewoverton) [inline components/ActionSheet/src/MDCActionSheetController.h:51]: Did we deem [text field stuff](https://developer.apple.com/documentation/uikit/uialertcontroller/1620093-addtextfield) overly complex for a first pass? Or are we just not concerned about it for this? Now that I think about it I don't think I've ever seen a text field in an action sheet, just an alert... 🤔

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: I think Apple uses this to render the actions differently, I believe all of our actions would render the same. But, if the team feels differently I'd love know that.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h:51]: I think Action Sheet will just be a list of items.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (jverkoey) [inline components/ActionSheet/src/MDCActionSheetController.h]: Consider marking this API as subclassing_restricted as well until shown that we need it otherwise.

**Review by jverkoey — APPROVED** [review]
LGTM!

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h:64]: Should there be a property to allow clients to set the dismiss/collapse button behavior in the header? From the [Material Guidelines article for Bottom Sheets](https://material.io/design/components/sheets-bottom.html#behavior), it seems like sometimes they can be dismissed (X) and other times they are just collapsed back to half (v).

Mock screenshots showing the two header icons side-by-side:
bottomsheet-dismiss-affordance-variety

bottomsheet-modal-dismiss-affordance

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h:134]: Should there be APIs for theming? From the [Material Guidelines theming section](https://material.io/design/components/sheets-bottom.html#theming) it looks like color and shape (not yet built on iOS) theming are permitted.

Suggestion:
```objective
@property(nonatomic, strong) UIColor *titleTextColor;
@property(nonatomic, strong) UIFont *titleTextFont;
@property(nonatomic, strong) UIColor *messageTextColor;
@property(nonatomic, strong) UIFont *messageTextFont;
@property(nonatomic, strong) UIColor *actionTextColor;
@property(nonatomic, strong) UIColor *actionTintColor; // for template image tinting
@property(nonatomic, strong) UIColor *inkColor; // Overlay ripple color for touch feedback
```

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: Nit/spelling: s/UIConnectSizeCategoryAdjusting/UIContentSizeCategoryAdjusting/

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: What's the idea for a nullable block here? Do we have an expected use case for an action sheet with buttons that do nothing?

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: Can the image be nullable? Are images required for action items from a design perspective or only recommended? The Design guidelines article seems to indicate that a Menu is a viable alternative on Mobile and the Google App on iOS uses menus without icons.

google_app_menu

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: Nit: Mark this as copy?

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (romoore) [reply components/ActionSheet/src/MDCActionSheetController.h]: I agree with @codeman7 here.  Since modal bottom sheets with action lists aren't expressly intended for confirmation of actions ([like a Dialog is](https://material.io/design/components/dialogs.html#usage)), I don't think we need to mark destructive/cancel actions. I would expect an action like "Delete" would result in a confirmation message or would support undo.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (romoore) [inline components/ActionSheet/src/MDCActionSheetController.h]: Nit: s/UIAlertControllerSytleActionSheet/UIAlertControllerStyleActionSheet/

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (romoore) [reply components/ActionSheet/src/MDCActionSheetController.h]: I think that would cause problems since those parameters are currently marked `nonnull` and Swifters could accidentally unwrap some non-optional nils.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h:134]: I am going to add all of those API's at a later date.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Wouldn't that make a copy of the array and therefore possibly introduce bugs if this array changes and the underlying array doesn't reflect those changes?

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: I don't know why a client would have a button that does nothing but I think it is a possibility. I don't see a benefit in marking it as nonnull.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h:64]: Pushing this back to a later date.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (romoore) [reply components/ActionSheet/src/MDCActionSheetController.h]: When assigning an object from a class cluster that includes mutable variants, I believe the recommended practice is to use `copy` semantics. That prevents a caller from assigning a mutable object and changing it later.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (romoore) [reply components/ActionSheet/src/MDCActionSheetController.h]: I think the benefit is to help prevent programmer errors. Unless we have a use case, I would think attaching an action that does nothing to an action sheet should be considered incorrect.

Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: What about an action that "dismisses" the sheet?

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h]: Done

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (romoore) [reply components/ActionSheet/src/MDCActionSheetController.h]: That should be provided in the sheet's app bar (guidelines excerpt below).  Even if there were an explicit Dismiss/Cancel option, I would expect that is has an action to dismiss the action sheet.  Are you thinking that the ActionSheet would dismiss itself if any list item without an attached handler were activated?

> Full-screen modal bottom sheets contain a dismiss icon that dismisses the sheet entirely from view.
> &mdash; https://material.io/design/components/sheets-bottom.html#modal-bottom-sheet

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (randallli) [inline components/ActionSheet/src/MDCActionSheetController.h:112]: If the title and message are nullable why do we have a new designated initializer? 

Either make the title nonnull or allow the other initializers.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (randallli) [inline components/ActionSheet/src/MDCActionSheetController.h:158]: lets mark this designated init.
and make init unavailable. This will make sure the title will always be nonnull.

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (randallli) [inline components/ActionSheet/src/MDCActionSheetController.h:51]: Suggestion:
If we make the public API a protocol we can land this PR without the implementation.

This will make it easier to land smaller PRs and not have to wait until the component is complete before we land anything/everything at once. It also blends nicely with "use more protocols" advice out of our internal conference.

**Review by randallli — CHANGES_REQUESTED** [review]

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (galiak11) [inline components/ActionSheet/src/MDCActionSheetController.h:56]: Is this an alert controller or an action sheet controller? or both?
Are we considering combining action sheets and alerts in the same component - similar to UIAlertController?

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (galiak11) [inline components/ActionSheet/src/MDCActionSheetController.h]: only button(s)? what about title & message?

File: components/ActionSheet/src/MDCActionSheetController.h
Comment (codeman7) [reply components/ActionSheet/src/MDCActionSheetController.h:56]: It's just an action sheet controller. I don't think they will ever combine, I think we plan to make them individual components.

**Pull request closed without merging**
